### ISSUE TITLE
Flag pages that are dead/removed

### ISSUE THREAD

**Issue by Mr0grog** [edgi-govdata-archiving/web-monitoring#122]
One really useful feature for analysts would be to flag pages that have been removed. There’s no objective or even canonical way to monitor this (after all, we’ve seen plenty of servers that respond erroneously with 404s when they are under heavy load and unable to return a page that exists, as well as intermittent 404s when a page disappears for only a few days). So for now we are thinking:

A removed page is one that has responded with a 404 continuously for 14 days.

How might we measure this?

1. A script that repeatedly checks Wayback on a regular schedule to get the status of all the snapshots for a page over the past 14 days. Similar to the generalized ETL script we’re currently working on. We’d run it on a regular schedule, but probably not daily since it’s a big query. This would also require adding the ability to POST updates to a page in -db.

2. Seeing no new, different versions added to -db over a 14 day period when the most recent version was a 404. -db could easily manage its own async job to check and update this on a regular schedule.

Neither of these is perfect — (1) would be reliable, but only for URLs that we have Wayback checking for us on a regular basis. For other URLs, all bets are off. (2) Would be *much* easier to implement and maintain, but doesn’t give us a way to distinguish between pages that are persistently 404s and pages that our sources (like Versionista, Wayback, and Walk) have stopped monitoring and whose last snapshot was a 404 (because a page could always come *back,* but if we’ve stopped monitoring, we won’t know that).

I’d tend towards (2). Even with its imperfections, I think it’s probably good enough, is way lower impact to measure, and way lower effort to implement. In the future we could even make it more reliable by making `HEAD` requests for the pages we think are removed to make sure they are still 404s. (That said, depending how Walk works out, it might make doing this easier.)

@danielballan @jsnshrmn any thoughts here?

**Comment by Mr0grog** [comment]
Useful addendum from @ericnost here — some pages are also persistent 403, e.g. https://www.epa.gov/ghgreporting/subpart-w-basic-information, and that would be nice to flag, too.

Not sure on the best way to handle these needs, but some thoughts:

- A page-level status that follows the same rules (200 unless we are getting some 4xx or 5xx repeatedly for 14 days)
- A page-level status that lists the most recent status, plus the `removed` boolean flag
- Multiple flags, e.g. `removed` and `limited_access` and others as we come up with them
- Something else?

**Comment by Mr0grog** [comment]
Had a good conversation with @danielballan about this just now. We talked through a variety of solutions, but where we landed:

1. We should add a `status` field to `Page` objects and determine it by a two-step process (calculated over a window of the last N days):
    1. Is the status code *successful* (i.e. < 400) more than X% of the time? If yes, mark the page’s status as 200. (X should be a configurable threshold; we think this should probably be 50%-80%, but not sure.) *(To do: file issue and link it here)*
    2. Otherwise, find the the most common error status code (i.e. >= 400) during the time window and use that as the page’s `status`. *(To do: file issue and link it here)*

2. It’s probably important to find a more detailed way to display that information on the page details view in the UI. In addition to the above status code, we should provide some kind of histogram, graph, red/yellow/green confidence, or something to explain whether a page is constantly flipping, or whether its status codes have otherwise varied a lot.

In the mean time, I need to file issues for the above and run some queries to try and get an idea for what the right success/error threshold is (1.1 above).
